verilog/formatting/formatter.cc

```diff
@@ -230,15 +230,265 @@ static void DeterminePartitionExpansion(partition_node_type* node,
+//
+// 2. Prefers first one (same number of lines, prefers unwrapped result)
+// pretend column limit here                           ->|
+// [function fffffffffff(type_a aaaa, type_b bbbbb,]
+// [>>>>>>>>>>>>>>>>>>>>>type_c cccccc, type_d dddddddd,]
+// [>>>>>>>>>>>>>>>>>>>>>type_e eeeeeeee, type_f ffff);]
+//
+// pretend column limit here                           ->|
+// [function fffffffffff(]
+// [>>>>type_a aaaa, type_b bbbbb, type_c cccccc,]
+// [>>>>type_d dddddddd, type_e eeeeeeee, type_f ffff);]
+//
+static void ReshapeFittingSubpartitions(partition_node_type& node,
+                                        const FormatStyle& style) {
+  auto& node_view = node.Value();
+  auto& children = node.Children();
+  const auto& uwline = node_view.Value();
+  const auto partition_policy = uwline.PartitionPolicy();
+
+  // Handle only kAppendFittingSubpartitions policy, ignore rest of them
+  switch (partition_policy) {
+  case PartitionPolicyEnum::kAppendFittingSubPartitions:
+    break;
+
+  default:
+    return;
+  }
+
+  // Check if this partition fits in line
+  // If so, Unexpand() and don't waste time trying to fit it
+  // If not, force Expand()
+  if (verible::FitsOnLine(uwline, style).fits) {
+    node_view.Unexpand();
+    return ;
+  } else {
+    // if not, force expansion and apply alogrithm
+    node_view.Expand();
+  }
+
+  //
+  // One-level partition tree expansion
+  // (ignores subpartition policies)
+  //
+  // e.g. input:
+  // -----------
+  // function fffffffffff(type_a aaaa, type_b bbbbb);
+  //
+  // partition tree:
+  // ---------------
+  // { ([<auto>]) @{2}, policy: append-fitting-sub-partitions
+  //   { ([function fffffffffff (]) }
+  //   { (>>>>[<auto>]) @{2,1}, policy: fit-else-expand
+  //     { (>>>>[type_a aaaa ,]) }
+  //     { (>>>>[type_b bbbbb ,]) }
+  //   }
+  //   { ([) ;]) }
+  // }
+  //
+  // output:
+  // -------
+  // [function fffffffffff (]
+  // >>>>[type_a aaaa ,])
+  // >>>>[type_b bbbbb])
+  // [) ;]
+  //
+  std::vector<UnwrappedLine> unwrapped_lines;
+  for (const auto& itr : children) {
+    if (itr.Children().size()) {
+      for (const auto& chld : itr.Children()) {
+        unwrapped_lines.push_back(chld.Value().Value());
+      }
+    } else {
+      unwrapped_lines.push_back(itr.Value().Value());
+    }
+  }
+
+  const std::vector<UnwrappedLine> *fitted_lines = nullptr;
+  std::vector<UnwrappedLine> unwrapped;
+  std::vector<UnwrappedLine> wrapped;
+
+  // Format unwrapped_lines. At first without forced wrap after first line
+  bool wrapped_first_token = AppendFittingSubpartitions(unwrapped, unwrapped_lines, style, false);
+
+  if (wrapped_first_token) {
+    // First token was forced to wrap so there's no need to
+    // generate wrapped version (it's already wrapped)
+    fitted_lines = &unwrapped;
+  } else {
+    // Generate wrapped version to compare results
+    AppendFittingSubpartitions(wrapped, unwrapped_lines, style, true);
+
+    // Compare number of fitted lines.
```

Comment:
What happens if among those lines, there was a single unwrapped line that did not fit and is expected to need to be wrapped by line_wrap_searching?  It's unknown at this time how many (vertical) lines it would have actually taken after wrapping.  This is something to think about for later.  In the long run, I am interested in exploring dynamic programming approaches to building up subproblems' solutions to do a more holistic optimization over the space of subpartition formattings.